balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,629 @@
+		serializerCancel: cancel,
+		state:            connectivity.Connecting,
+	}
+	b.logger = internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(logPrefix, b))
+	return b
+}
+
+func (b pickfirstBuilder) Name() string {
+	return Name
+}
+
+func (pickfirstBuilder) ParseConfig(js json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	var cfg pfConfig
+	if err := json.Unmarshal(js, &cfg); err != nil {
+		return nil, fmt.Errorf("pickfirst: unable to unmarshal LB policy config: %s, error: %v", string(js), err)
+	}
+	return cfg, nil
+}
+
+type pfConfig struct {
+	serviceconfig.LoadBalancingConfig `json:"-"`
+
+	// If set to true, instructs the LB policy to shuffle the order of the list
+	// of endpoints received from the name resolver before attempting to
+	// connect to them.
+	ShuffleAddressList bool `json:"shuffleAddressList"`
+}
+
+// scData keeps track of the current state of the subConn.
+type scData struct {
+	subConn balancer.SubConn
+	state   connectivity.State
+	addr    resolver.Address
+	lastErr error
+}
+
+func newSCData(b *pickfirstBalancer, addr resolver.Address) (*scData, error) {
+	sd := &scData{
+		state: connectivity.Idle,
+		addr:  addr,
+	}
+	sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
+		StateListener: func(state balancer.SubConnState) {
+			// Store the state and delegate.
+			b.serializer.TrySchedule(func(_ context.Context) {
+				sd.state = state.ConnectivityState
+				b.updateSubConnState(sd, state)
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	sd.subConn = sc
+	return sd, nil
+}
+
+type pickfirstBalancer struct {
+	// The following fields are initialized at build time and read-only after
+	// that and therefore do not need to be guarded by a mutex.
+	logger *internalgrpclog.PrefixLogger
+	cc     balancer.ClientConn
+
+	// The serializer and its cancel func are initialized at build time, and the
+	// rest of the fields here are only accessed from serializer callbacks (or
+	// from balancer.Balancer methods, which themselves are guaranteed to be
+	// mutually exclusive) and hence do not need to be guarded by a mutex.
+	// The serializer is used to ensure synchronization of updates triggered
+	// from the idle picker and the already serialized resolver,
+	// subconn state updates.
+	serializer       *grpcsync.CallbackSerializer
+	serializerCancel func()
+	state            connectivity.State
+	subConns         *resolver.AddressMap // scData for active subonns mapped by address.
+	addressList      addressList
+	firstPass        bool
+}
+
+func (b *pickfirstBalancer) ResolverError(err error) {
+	b.serializer.TrySchedule(func(_ context.Context) {
+		b.resolverError(err)
+	})
+}
+
+func (b *pickfirstBalancer) resolverError(err error) {
+	if b.logger.V(2) {
+		b.logger.Infof("Received error from the name resolver: %v", err)
+	}
+	if b.state == connectivity.Shutdown {
+		return
+	}
+	// The picker will not change since the balancer does not currently
+	// report an error.
```

Comment:
> Java unconditionally puts the LB in TF

That doesn't seem like the correct thing to do. But it also looks like they are invoking this method only when NR returns zero addresses. And if that is the case, then the behavior seems correct to me.

We handle the case where the name resolver returns zero addresses, and the case where the name resolver returns an error in the same method `resolverError` which doesn't lend itself to doing the right thing under all circumstances looks like.

To me, this is what the behavior should be:
- If we have a previous good update from the NR (i.e. we are currently `READY`), and we get an error from it, ignore the error and continue to use the old good state. For all other cases where we get an error from the name resolver, move to `TF`
- If we get zero addresses from the NR, move to TF unconditionally

@dfawley : Thoughts?

Reply:
> To me, this is what the behavior should be:

That sounds right to me.  And I think it's handled by the old PF by that first `if subConn == nil` which is the "has no addresses already".

However, I don't think I agree with:

> doesn't lend itself to doing the right thing under all circumstances

In what situation do you think we're doing it wrong?


Also, as an aside, we should be handling `ResolverError`s synchronously just like `UpdateClientConnState` calls.

Reply:
@easwars, to ensure point 2 (`If we get zero addresses from the NR, move to TF unconditionally`) is correctly handled, I have a check in `updateClientConnState` for an empty address list which sets `balancer.sate = TF` before calling `resolverError`.
```go
	if len(state.ResolverState.Addresses) == 0 && len(state.ResolverState.Endpoints) == 0 {
		// Cleanup state pertaining to the previous resolver state.
		// Treat an empty address list like an error by calling b.ResolverError.
		b.state = connectivity.TransientFailure
		b.resolverError(errors.New("produced zero addresses"))
		return balancer.ErrBadResolverState
	}
```
This should be equivalent to the check for `if subConn == nil` and setting `b.state = TF` in the old PF's `ResolveError()`.

Reply:
@dfawley changed `ResolveError` to be synchronous.

Reply:
> That sounds right to me. And I think it's handled by the old PF by that first if subConn == nil which is the "has no addresses already".

That's true.

Reply:
>> doesn't lend itself to doing the right thing under all circumstances

> In what situation do you think we're doing it wrong?

Hmm, I feel more than correctness, its about how easy it is to follow all the possible scenarios. Looks like the second bullet is handled in the old pickfirst and new one. But the first bullet is completely handled only in the old pickfirst. It is not handled in the new pickfirst. Specifically the case where the NR has not sent an update so far, but reports an error. The new pickfirst will see that `	if b.state != connectivity.TransientFailure { ... }` and will return early instead of moving to TF.

Can we also have a test for this scenario?

Reply:
In Go, if we get an error from the NR before any update, we try to apply the default service config (if one exists). https://github.com/grpc/grpc-go/blob/master/clientconn.go#L734-L745

But if we haven't built the LB policy yet, the gracefulswitch balancer handles this case and puts the channel in TF. So, looks like the expected behavior happens, but it is quite convoluted. 

Maybe, we have a comment in the new pickfirst's resolverError() method saying that if the NR returns an error before sending an update, it is handled by the gracefulswitch balancer (which is always the top-level LB on any channel).

Reply:
> Can we also have a test for this scenario?

Added a unit test that verifies that the ClientConn goes from `IDLE->TF->Shutdown` when a resolver error is reported and the clientconn is shutdown.

> we have a comment in the new pickfirst's resolverError() 

Added a comment as suggested.